src/Streams.sol

```diff
@@ -144,6 +144,59 @@ library StreamConfigImpl {
     }
 }
 
+/// @notice The list of 8 hints for max end time calculation.
+/// They are constructed as a concatenation of 8 32-bit values:
+/// `the leftmost hint (32 bits) | ... | the rightmost hint (32 bits)
+type MaxEndHints is uint256;
+
+using MaxEndHintsImpl for MaxEndHints global;
+
+library MaxEndHintsImpl {
+    /// @notice Create a list of 8 zero value hints.
+    /// @return hints The list of hints.
+    // slither-disable-next-line dead-code
+    function create() internal pure returns (MaxEndHints hints) {
+        return MaxEndHints.wrap(0);
+    }
+
+    /// @notice Add a hint to the list of hints as the rightmost and remove the leftmost.
+    /// @param hints The list of hints.
+    /// @param hint The added hint.
+    /// @return newHints The modified list of hints.
+    // slither-disable-next-line dead-code
+    function push(MaxEndHints hints, uint32 hint) internal pure returns (MaxEndHints newHints) {
+        // `hints` has value:
+        // `leftmost hint (32 bits) | other hints (224 bits)`
+        // By bit shifting we get value:
+        // `other hints (224 bits) | zeros (32 bits)`
+        // By bit masking we get value:
+        // `other hints (224 bits) | rightmost hint (32 bits)`
+        return MaxEndHints.wrap((MaxEndHints.unwrap(hints) << 32) | hint);
+    }
+
+    /// @notice Remove and return the rightmost hint, and add the zero value hint as the leftmost.
+    /// @param hints The list of hints.
+    /// @return newHints The modified list of hints.
+    /// @return hint The removed hint.
+    function pop(MaxEndHints hints) internal pure returns (MaxEndHints newHints, uint32 hint) {
```

Comment:
It makes sense to implement it with a `pop` instead of a `read` operation for a specific list index. (more costly)